core/src/Cache/CacheFallbackOperations.cs

```diff
@@ -238,29 +237,46 @@ internal class CacheFallbackOperations
         }
 
         public static AdalResultWrapper FindMsalEntryForAdal(ITokenCacheAccessor tokenCacheAccessor, string authority,
-            string clientId, string upn)
+            string clientId, string upn, RequestContext requestContext)
         {
-            foreach (var rtString in tokenCacheAccessor.GetAllRefreshTokensAsString())
-            {
-                var rtCacheItem =
-                    JsonHelper.DeserializeFromJson<MsalRefreshTokenCacheItem>(rtString);
+            var environment = new Uri(authority).Host;
 
-                var accountStr = tokenCacheAccessor.GetAccount(rtCacheItem.GetAccountItemKey());
-
-                var accountItem =
-                    JsonHelper.DeserializeFromJson<MsalAccountCacheItem>(accountStr);
+            List<MsalAccountCacheItem> accounts = new List<MsalAccountCacheItem>();
+            foreach (string accountStr in tokenCacheAccessor.GetAllAccountsAsString())
+            {
+                var accountItem = JsonHelper.TryToDeserializeFromJson<MsalAccountCacheItem>(accountStr, requestContext);
+                if (accountItem != null && accountItem.Environment.Equals(environment))
+                {
+                    accounts.Add(accountItem);
+                }
+            }
+            if (accounts.Count > 0)
+            {
+                foreach (var rtString in tokenCacheAccessor.GetAllRefreshTokensAsString())
+                {
+                    var rtCacheItem =
+                        JsonHelper.TryToDeserializeFromJson<MsalRefreshTokenCacheItem>(rtString, requestContext);
 
-                //TODO - authority check needs to be updated for alias check
-                if (new Uri(authority).Host.Equals(rtCacheItem.Environment) && rtCacheItem.ClientId.Equals(clientId) &&
-                    accountItem.PreferredUsername.Equals(upn))
-                    return new AdalResultWrapper
+                    //TODO - authority check needs to be updated for alias check
+                    if (rtCacheItem != null && environment.Equals(rtCacheItem.Environment)
+                        && rtCacheItem.ClientId.Equals(clientId))
                     {
-                        Result = new AdalResult(null, null, DateTimeOffset.MinValue),
-                        RefreshToken = rtCacheItem.Secret,
-                        RawClientInfo = rtCacheItem.RawClientInfo
-                    };
+                        foreach (MsalAccountCacheItem accountCacheItem in accounts)
+                        {
```

Comment:
logic seems very complicated. I would like to see a comment explaining why we have to to this.

Reply:
added

Reply:
I do not see the comments yet.

---
In reply to: [199976862](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1108#discussion_r199976862) [](ancestors = 199976862)

Reply:
I suppose I still don't get it... would have preferred additional methods which could have helped clarified the intent of the code

---
In reply to: [200183263](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1108#discussion_r200183263) [](ancestors = 200183263,199976862)